kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopCredentialsManager.scala

```diff
@@ -219,13 +233,23 @@ class HadoopCredentialsManager private (name: String) extends AbstractService(na
                 s" $renewalRetryWait ms",
               e)
             scheduleRenewal(userRef, renewalRetryWait)
+            // throw exception when one-time execution fails,
+            // so that client side can be aware of this
+            if (waitCompletion) {
+              throw new KyuubiException(s"One-time execution failed for token update task " +
```

Comment:
I'm confused with the code, we always do `scheduleRenewal` if encounter Exception, so how can we go into here ? But I see the test has checked the `KyuubiException` successfully, am I miss something ? @turboFei

Reply:
@ulysses-you `scheduleRenewal` only submits a task. It does not do recursive invocation.

Reply:
it seems if it is waitCompletion, we do not need `scheduleRenewal` again

Reply:
@zhouyifan279 @ulysses-you, IMO, `waitCompletion` doesn't make sense to me. If this param is used to throw the exception, the `waitCompletion` is unncessary.

Reply:
> it seems if it is waitCompletion, we do not need scheduleRenewal again

@ulysses-you  we need scheduleRenewal again so that the credentials can keep renew itself, otherwise, it might expire later.

Reply:
@SteNicholas 

Even we wait the renew task completion this time, we also need to renew the credentials in the future.

We throw exception because that we need let the client side be aware of this exception.

Reply:
@turboFei , do you mean the client side uses `waitCompletion` to  be aware of this exception? This way doesn't look so good to me.

Reply:
> do you mean the client side uses waitCompletion to be aware of this exception? 

No, it is not controlled by client side, we just log error message here so that client side can fetch the log info.
Before, if there is failure when renewing credentials, nothing will be logged.
But I think we should let customer be aware of this, because for the corner case, the follow operations might fail, the error message can provide some insight for customer.

Reply:
@SteNicholas 
We can follow this comment: https://github.com/apache/incubator-kyuubi/pull/2059#discussion_r822279846

Add a new method to wait creds ready.